pkg/provider/naming.go

```diff
@@ -0,0 +1,44 @@
+package provider
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/pulumi/pulumi/pkg/tokens"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+var renameBlacklist = sets.NewString("v1/Namespace")
+
+// assignName generates a name for an object if it doesn't already have one. Uses DNS-1123-compliant
+// characters.
+func assignName(obj *unstructured.Unstructured, base tokens.QName) {
+	if !renameBlacklist.Has(obj.GroupVersionKind().String()) {
+		replaceNameIfUnnamed(obj, fmt.Sprintf("%s-%s", base, randString(8)))
+	} else {
+		replaceNameIfUnnamed(obj, string(base))
+	}
+}
+
+// replaceNameIfUnnamed will replace an object's name if it doesn't already have one.
+func replaceNameIfUnnamed(obj *unstructured.Unstructured, name string) {
+	if obj.GetName() == "" {
+		meta := obj.Object["metadata"]
+		if meta == nil {
+			obj.Object["metadata"] = map[string]interface{}{"name": name}
+		} else {
+			meta.(map[string]interface{})["name"] = name
+		}
+	}
+}
+
+func randString(n int) string {
+	var alphabet = []rune("abcdefghijklmnopqrstuvwxyz0123456789")
+
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = alphabet[rand.Intn(len(alphabet))]
```

Comment:
Note that the default `rand` source is seeded with the same value each time the process starts, so this will be essentially deterministic. If that matters, you should seed the source in this package's `init` function.